sys/linux/auto.txt

```diff
@@ -242,18 +247,18 @@ fchmodat2$auto(dfd fd_dir, filename ptr[in, filename], mode int16, flags int32)
 fchown$auto(fd fd, user uid, group gid)
 fchown32$auto(fd fd, user uid, group gid)
 fchownat$auto(dfd fd_dir, filename ptr[in, filename], user uid, group gid, flag int32)
-fcntl$auto(fd fd, cmd int32, arg intptr)
+fcntl$auto(fd fd, cmd int32, arg intptr) fd
```

Comment:
fcntl() is a multi-purpose call, do we always want to treat its return value as an fd?

Reply:
"multi-purpose" is a vague term. E.g. if mmap a multi-purpose? It can mmap anon memory, or an fd (and fd's of different types). I guess all syscalls can be considered multi-purpose to some degree. Currently we don't have a notion of "multi-purpose" in type inference for simplicity. There is a flow from allocfd, so we say it's an fd.

At some we probably want some notion of multi-purpose. But it's a more complex problem. First, we will need to infer what's a "multi-purpose", how exactly this multi-purposedness is controlled. And then do context-sensitive data flow analysis based on that discriminating factor.

There are these TODOs for this:

```
// - Add context-sensitivity at least on switched arguments (ioctl commands).
// - Infer other switched arguments besides ioctl commands.
```